File: identhub/src/main/java/de/solarisbank/identhub/domain/contract/ConfirmContractSignUseCaseFactory.java
package de.solarisbank.identhub.domain.contract;

import de.solarisbank.sdk.core.di.internal.Factory;
import de.solarisbank.sdk.core.di.internal.Preconditions;
import de.solarisbank.sdk.core.di.internal.Provider;

public class ConfirmContractSignUseCaseFactory implements Factory<ConfirmContractSignUseCase> {

    private final Provider<ContractSignRepository> contractSignRepositoryProvider;

    public ConfirmContractSignUseCaseFactory(Provider<ContractSignRepository> contractSignRepositoryProvider) {
        this.contractSignRepositoryProvider = contractSignRepositoryProvider;
    }

    public static ConfirmContractSignUseCaseFactory create(Provider<ContractSignRepository> contractSignRepositoryProvider) {
        return new ConfirmContractSignUseCaseFactory(contractSignRepositoryProvider);
    }

    @Override
    public ConfirmContractSignUseCase get() {
        return Preconditions.checkNotNull(
                new ConfirmContractSignUseCase(contractSignRepositoryProvider.get()),
                "Cannot return null from provider method"
        );
    }
}

File: identhub/src/main/java/de/solarisbank/identhub/domain/di/contract/ConfirmContractSignUseCaseFactory.kt
package de.solarisbank.identhub.domain.di.contract

import de.solarisbank.identhub.data.contract.ContractSignRepository
import de.solarisbank.identhub.data.contract.step.parameters.QesStepParametersRepository
import de.solarisbank.identhub.domain.contract.ConfirmContractSignUseCase
import de.solarisbank.sdk.domain.usecase.IdentificationPollingStatusUseCase
import de.solarisbank.sdk.feature.di.internal.Factory
import de.solarisbank.sdk.feature.di.internal.Preconditions
import de.solarisbank.sdk.feature.di.internal.Provider

class ConfirmContractSignUseCaseFactory(
    private val contractSignRepositoryProvider: Provider<ContractSignRepository>,
    private val identificationPollingStatusUseCaseProvider: Provider<IdentificationPollingStatusUseCase>,
    private val qesStepParametersRepositoryProvider: Provider<QesStepParametersRepository>
) : Factory<ConfirmContractSignUseCase> {
    override fun get(): ConfirmContractSignUseCase {
        return Preconditions.checkNotNull(
            ConfirmContractSignUseCase(
                contractSignRepositoryProvider.get(),
                identificationPollingStatusUseCaseProvider.get(),
                qesStepParametersRepositoryProvider.get()
            ),
            "Cannot return null from provider method"
        )
    }

    companion object {
        @JvmStatic
        fun create(
            contractSignRepositoryProvider: Provider<ContractSignRepository>,
            identificationPollingStatusUseCaseProvider: Provider<IdentificationPollingStatusUseCase>,
            qesStepParametersRepositoryProvider: Provider<QesStepParametersRepository>
        ): ConfirmContractSignUseCaseFactory {
            return ConfirmContractSignUseCaseFactory(
                contractSignRepositoryProvider,
                identificationPollingStatusUseCaseProvider,
                qesStepParametersRepositoryProvider
            )
        }
    }
}